jax/_src/lax/search.py

```diff
@@ -0,0 +1,179 @@
+
+  Args:
+    sorted_arr : N-dimensional array, which is assumed to be sorted in increasing
+      order along ``dimension``.
+    query : N-dimensional array of query values.
+    side : {'left', 'right'}. If 'left', find the index of the first suitable
+      location. If 'right', find the index of the last.
+    dimension : integer specifying the dimension along which to insert query values.
+    batch_dims : integer specifying the number of leading dimensions of `sorted_arr`
+      and `query` to treat as batch dimensions.
+    method : {'default', 'scan', 'sort'} The method used to compute the result.
+      Outside JIT, this defaults to "scan". Within JIT, this defaults to "scan"
+      on CPU, and "sort" otherwise. Assume ``M = query.size`` and
+      ``N = sorted_arr.shape[dimension]``. ``method='scan'`` uses an *O[M log N]*
+      binary search algorithm, while ``method='sort'`` uses an *O[(M+N) log(M+N)]*
+      co-sorting approach. Although ``method='scan'`` may appear to have better
+      scaling overall, in practice it cannot take advantage of parallelism inherent
+      to accelerators, and so in most cases is slower than ``method='sort'`` on GPU
+      and TPU backends.
+
+  Returns:
+    indices : an array specifying the insertion locations of `query` into `sorted_arr`.
+  """
+  dimension = util.canonicalize_axis(dimension, len(sorted_arr.shape))
+  batch_dims = core.concrete_or_error(operator.index, batch_dims, context="searchsorted batch_dims argument")
+  return searchsorted_p.bind(sorted_arr, query, side=side, dimension=dimension, batch_dims=batch_dims, method=method)
+
+def _searchsorted_abstract_eval(sorted_arr, query, *, side, dimension, batch_dims, method):
+  batch_dims = operator.index(batch_dims)
+  if batch_dims < 0:
+    raise ValueError(f"batch_dims must be a non-negative integer; got {batch_dims}")
+  if sorted_arr.dtype != query.dtype:
+    raise ValueError("dtypes of sorted_arr and query must match; got "
+                     f"{sorted_arr.dtype} and {query.dtype}")
+  if method not in ["default", "scan", "sort"]:
+    raise ValueError(f"invalid argument method={method!r}; expected 'default', 'scan', or 'sort'.")
+  if side not in ["left", "right"]:
+    raise ValueError(f"invalid argument side={side!r}, expected 'left' or 'right'")
+  if not batch_dims <= dimension < sorted_arr.ndim:
+    raise ValueError(f"dimension={dimension} must be in range [{batch_dims}, {sorted_arr.ndim})")
+  if sorted_arr.shape[:batch_dims] != query.shape[:batch_dims]:
+    raise ValueError(f"batch dimension sizes must match; got {sorted_arr.shape[:batch_dims]} != "
+                     f"{query.shape[:batch_dims]}")
+  shape = (*sorted_arr.shape[:batch_dims],
+           *(s for d, s in enumerate(sorted_arr.shape) if d >= batch_dims and d != dimension),
+           *(s for d, s in enumerate(query.shape) if d >= batch_dims))
+  dtype = np.int32 if sorted_arr.shape[dimension] < np.iinfo(np.int32).max else np.int64
```

Comment:
Considering the sort based approach, it should be the length of the concatenated array that is smaller than `1 << 31` right?

Reply:
Good catch - done.